Vladimir Smirnov

Dodo888 opened the pull request:
Tests for library flow.js

Dodo888 committed:
Apple Pie

Test for library flow.js

Zhigalov commented on lib/flow.js:
```diff
@@ -0,0 +1,68 @@
+'use strict';
```
Чтобы не дублировать решение задачи в двух репозиториях - можно подключить его как npm  зависимость.
Сергей на последней паре рассказывал как можно указать git репозитории в качестве зависимости.

Zhigalov commented on test/flow-test.js:
```diff
@@ -0,0 +1,141 @@
+                callback(null, data+5)
+            }, function (data, callback) {
+                callback(null, data*2)
+            }],
```
Это действительно проверяет порядок запуска.
Но я советую посмотреть в сторону использования [spies](http://sinonjs.org/docs/#spies)
Например используя `spy.calledBefore(anotherSpy);` можно проверить порядок вызова.
В таком случае тест будет более очевидным.

Dodo888 committed:
Banana Breed

-Flow library is now a dependency
-Better testing by sinon.spies
-Fluent language by sinon-chai

Dodo888 commented:
Добавил спаев и зависимость)

Zhigalov commented on package.json:
```diff
@@ -0,0 +1,28 @@
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Dodo888/webdev-tasks-3.git"
```
:+1:

Zhigalov commented on test/flow-test.js:
```diff
@@ -0,0 +1,160 @@
+const mocha = require('mocha');
+const chai = require('chai')
+    , expect = chai.expect
+    , should = chai.should();
```
`should` в переменной кажется тебе нигде не нужен. Можно не писать

Zhigalov commented on test/flow-test.js:
```diff
@@ -0,0 +1,160 @@
+const flow = require('flow/lib/flow.js');
```
В `require` можно не писать расширение файла, он тебе сам сначала `.js` потом `.json` попытается подставить

Zhigalov commented on test/flow-test.js:
```diff
@@ -0,0 +1,160 @@
+const flow = require('flow/lib/flow.js');
+const assert = require('assert');
```
Кажется эта штука нигде не используется

Zhigalov commented on test/flow-test.js:
```diff
@@ -0,0 +1,160 @@
+                data.should.equal(14);
+                func1.should.be.calledBefore(func2);
+                func2.should.be.calledBefore(func3);
+            });
```
В целом все правильно, но предпологается что колбеки func1, func2, func3 будут асинхронными функциями.
То есть они будут использовать апи ноды, вставать в очередь событий и все такое. Если забыл - отсылаю тебя к лекции про асинхронность в прошлом семестре.
По факту хоть у тебя сейчас и есть колбеки, они синхронные. А на синхронных колбеках правильно параллельность не протестировать.
Самы простой способ эмулировать асинхронность - использовать  `setTimeout`.
Интересные вещи получаются если выставить разные времена, особенно в тастах на параллельный вызов.

Zhigalov commented on test/flow-test.js:
```diff
@@ -0,0 +1,160 @@
+            callback(null, 2);
+        });
+        var func2 = sinon.spy(function (data, callback) {
+            callback(null, data+5);
```
Зачем эта сложная математика?
Может лучше проверить отдельным тестом что все параметры правильно передаются?

Zhigalov commented on test/flow-test.js:
```diff
@@ -0,0 +1,160 @@
+        flow.parallel([func1, func2], function (error, data) {
+            data.should.be.a('array');
+            data.length.should.equal(2);
+            data[1].should.equal(2);
```
a `data[0]` почему не проверил?

Zhigalov commented on test/flow-test.js:
```diff
@@ -0,0 +1,160 @@
+            function (error, data) {
+                error.should.be.ok;
+            });
+    });
```
Я бы еще проверил что в случае ошибки остальные коллбеки доработали правильно

Zhigalov commented on test/flow-test.js:
```diff
@@ -0,0 +1,160 @@
+            func.withArgs(10).should.be.calledOnce;
+            func.withArgs(20).should.be.calledOnce;
+            func.withArgs(30).should.be.calledOnce;
+        });
```
Еще можно проверить что `func` позвали ровно три раза

Zhigalov commented:
Уже очень хорошо, кажется осталось совсем немного поправить

Dodo888 committed:
Cupcake

Asynchronous functions in tests
More test cases

Dodo888 commented:
Пофиксил)

Zhigalov commented on test/flow-test.js:
```diff
@@ -0,0 +1,240 @@
+                expect(error).not.to.be.ok;
+                func1.should.be.calledBefore(func2);
+                func2.should.be.calledBefore(func3);
+            });
```
Когда тест стал асинхронным нужно дожидаться когда он выполнится, для этого нужно использовать `done`
Сейчас тест ложноположительный, потому что добавив например `func3.should.be.calledBefore(func1);` он по прежнему будет проходить

Dodo888 committed:
Donut

Added done to aynchronous functions

Dodo888 commented:
Добавил done в асинхронные тесты

Zhigalov commented:
Все правильно
